### PR TITLE
Allow negative min_qty

### PR DESCRIPTION
**Pull request by aleron75**
This WIP PR addresses issue #1370 to allow negative min_qty (aka out-of-stock threshold).

- Add around plugin on `\Magento\CatalogInventory\Model\System\Config\Backend\Minqty::beforeSave()` to allow `min_qty` to be assigned a value below 0
- Add around plugin on `\Magento\Catalog\Controller\Adminhtml\Product\Initialization\StockDataFilter::filter()` to allow `min_qty` to be assigned a value below 0
- Add around plugin on `\Magento\CatalogInventory\Api\Data\StockItemInterface::getMinQty()` to correctly calculate `min_qty` value
- Update salable conditions related to backorders in resource model of InventorySales

Still to do at the time of submitting the WIP PR:

- Check all caclulations (running test suite can be useful to the purpose)
- Check whether/how only-x-left functionality is affected
- Add proper explanation in Admin Panel of the meaning of how `min_qty` value according to backorders enabling

### CODE REVIEW

**Comment by smoskaluk** [comment]
@aleron75  Hi Alessandro, during my testing I could add to the cart any qty of product (limitation didn't work)

**Comment by aleron75** [comment]
Hi @smoskaluk, thank you for your feedback, I will investigate asap.
Cheers

**Comment by aleron75** [comment]
Hi @smoskaluk my last commit should fix the problem of adding any qty to the cart even if negative min_qty is set. 
This PR is still under development so any additional feedback is greatly appreciated.
Cheers

**Comment by smoskaluk** [comment]
@aleron75 Hi Alessandro! It looks like your last commit fixed problem with adding products to the cart. What I've found as well, that simple products couldn't be shipped afterwards. Only qty allocated on the sources. Virtual products may be invoiced  correctly - but if you do credit memo with return to the stock - then all qty adding to the source (real + value of out of stock threshold) 
All regards!

**Comment by smoskaluk** [comment]
@aleron75 One more thing I concern about - when you place order with backorders - you get negative qty on stock - so you got some indication - but with negative OOST that not happen.

**Comment by aleron75** [comment]
Hi @smoskaluk, we are addressing the last issues you refer to, I'll update here asap, thank you

File: app/code/Magento/InventoryCatalog/Plugin/Catalog/Controller/Adminhtml/Product/Initialization/StockDataFilter/AllowNegativeMinQtyPlugin.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+{
+    /**
+     * Allow min_qty to be assigned a value below 0.
+     * @param StockDataFilter $subject
+     * @param callable $proceed
+     * @param array $stockData
+     * @return mixed
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function aroundFilter(
```
Comment (maghamed) [inline app/code/Magento/InventoryCatalog/Plugin/Catalog/Controller/Adminhtml/Product/Initialization/StockDataFilter/AllowNegativeMinQtyPlugin.php]: you no need around plugin here, just use after plugin instead

File: app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/Api/Data/StockItemInterface/AdaptMinQtyToBackordersPlugin.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+use Magento\CatalogInventory\Api\Data\StockItemInterface;
+
+class AdaptMinQtyToBackordersPlugin
+{
+    /**
+     * @param StockItemInterface $subject
+     * @param callable $proceed
+     * @return int
+     */
+    public function aroundGetMinQty(StockItemInterface $subject, callable $proceed)
```
Comment (maghamed) [inline app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/Api/Data/StockItemInterface/AdaptMinQtyToBackordersPlugin.php]: you no need around plugin here, just use after plugin instead

**Review by maghamed — CHANGES_REQUESTED** [review]
Also , looks like we need to extend the CatalogInventory function 
```
    public function verifyStock(StockItemInterface $stockItem)
    {
        if ($stockItem->getQty() === null && $stockItem->getManageStock()) {
            return false;
        }
        if ($stockItem->getBackorders() == StockItemInterface::BACKORDERS_NO
            && $stockItem->getQty() <= $stockItem->getMinQty()
        ) {
            return false;
        }
        return true;
    }
```
which we use in MSI, and which we affected after our changes

File: app/code/Magento/InventorySales/Model/ResourceModel/IsStockItemSalableCondition/BackordersCondition.php [hunk 3]
```diff
@@ -39,6 +39,7 @@ public function execute(Select $select): string
         $condition = (1 === $globalBackorders)
             ? 'legacy_stock_item.use_config_backorders = 1'
             : 'legacy_stock_item.use_config_backorders = 0 AND legacy_stock_item.backorders = 1';
+        $condition .= ' AND (legacy_stock_item.min_qty >= 0 OR legacy_stock_item.qty > legacy_stock_item.min_qty)';
```
Comment (maghamed) [inline app/code/Magento/InventorySales/Model/ResourceModel/IsStockItemSalableCondition/BackordersCondition.php]: this additional condition looks excessive to me.

**Comment by jhruehl** [comment]
@aleron75  I don't get why you create plugins to fix an issue in the underlying Magento Opensource modules instead of fixing the problem at its root.
Can you please give me a valid explanation?

**Pull request merged**